### PR TITLE
dojson: add links between collections

### PR DESCRIPTION
**Pull request by tomaszgy**
* Adds links by recid between Experiments and Institutions,
Jobs and Experiments and HepNames and PhdAdvisors.

* Updates JSON schemas.

* Includes tests.

Signed-off-by: Tomasz Gargas <tomaszgy@gmail.com>

### CODE REVIEW

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 1]
```diff
@@ -29,7 +29,8 @@
 from dojson import utils
 
 from ..model import hepnames, hepnames2marc
-from ...utils import classify_rank, force_single_element, get_record_ref
+from ...utils import (classify_rank, force_single_element,
+                      get_record_ref, get_recid_from_ref)
```
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]:
```python
from ...utils import (
    classify_rank,
    force_single_element,
    get_record_ref,
    get_recid_from_ref,
)
```

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 2]
```diff
@@ -464,7 +469,9 @@ def advisors2marc(self, key, value):
     """The advisors for all types of degrees."""
     return {
         'a': value.get('name'),
-        'g': value.get('_degree_type')
+        'g': value.get('_degree_type'),
+        'x': get_recid_from_ref(value.get('record')),
+        'y': '1' if value.get('curated_relation') else None
```
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Are you sure this is not `else '0'`?

File: inspirehep/dojson/experiments/fields/bd1xx.py [hunk 3]
```diff
@@ -40,10 +40,15 @@
...
-            self['affiliation'].append(raw_affiliation)
+        self.setdefault('affiliations', [])
+        name = value.get('u')
+        recid = value.get('z')
+        record = get_record_ref(recid, 'institutions')
+        self['affiliations'].append({
+            'curated_relation': record is not None,
+            'name': name,
+            'record': record
+            })
```
Comment (jacquerie) [inline inspirehep/dojson/experiments/fields/bd1xx.py]: Spacing is much better in https://github.com/inspirehep/inspire-next/pull/1451/files#diff-8f850f5dde975afda8a95a3023d5005cR124.

File: inspirehep/modules/records/jsonschemas/records/experiments.json [hunk 4]
```diff
@@ -5,14 +5,27 @@
...
-            "type": "array"
-        },
-        "affiliation_record": {
-            "$ref": "elements/json_reference.json",
-            "title": "Institution record URI"
+        "affiliations": {
+            "items": {
+                "properties": {
+                    "curated_relation": {
+                        "title": "Is the relation curated?",
```
Comment (jacquerie) [inline inspirehep/modules/records/jsonschemas/records/experiments.json]: I don't think this `title` adds anything to the property name. In fact, IMO, most `title`s in schemas should go away.

File: tests/unit/dojson/test_dojson_experiments.py [hunk 5]
```diff
@@ -145,30 +145,65 @@ def test_experiment_names_and_affiliation_from_marcxml_119():
         '  <datafield tag="119" ind1=" " ind2=" ">'
         '    <subfield code="a">CERN-ALPHA</subfield>'
         '    <subfield code="u">CERN</subfield>'
+        '    <subfield code="z">902725</subfield>'
         '  </datafield>'
         '</record>'
-    )
+    ) # record/1108206
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_experiments.py]: Nit: one more space between `)` and `#`.

File: tests/unit/dojson/test_dojson_experiments.py [hunk 6]
```diff
@@ -145,30 +145,65 @@ def test_experiment_names_and_affiliation_from_marcxml_119():
...
+            'curated_relation': True,
+            'name': 'CERN',
+            'record': {
+                '$ref': 'http://localhost:5000/api/institutions/902725',
+            },
+        },
+    ]
+
+    assert result['affiliations'] == expected_affiliations
     assert result['experiment_names'][0]['title'] == 'CERN-ALPHA'
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_experiments.py]: Looks cleaner:
```python
assert result['experiment_names'] == [{'title': 'CERN-ALPHA'}]
assert result['affiliations'] == [
    {
        ...
```

File: tests/unit/dojson/test_dojson_experiments.py [hunk 7]
```diff
@@ -145,30 +145,65 @@ def test_experiment_names_and_affiliation_from_marcxml_119():
...
+        '  <datafield tag="119" ind1=" " ind2=" ">'
         '    <subfield code="u">Edinburgh U.</subfield>'
+        '    <subfield code="z">902787</subfield>'
+        '  </datafield>'
+        '  <datafield tag="119" ind1=" " ind2=" ">'
+        '    <subfield code="u">Swansea U.</subfield>'
         '  </datafield>'
         '</record>'
-    )
+    ) # record/1228417
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_experiments.py]: Same nit: one more space between `)` and `#`.

File: tests/unit/dojson/test_dojson_hepnames.py [hunk 8]
```diff
@@ -323,6 +323,11 @@ def test_advisors(marcxml_to_json, json_to_marc):
             json_to_marc['701'][0]['a'])
     assert (marcxml_to_json['advisors'][0]['_degree_type'] ==
             json_to_marc['701'][0]['g'])
+    assert (get_recid_from_ref(marcxml_to_json['advisors'][0]['record']) ==
+            json_to_marc['701'][0]['x'])
+    assert (
+        ('1' if marcxml_to_json['advisors'][0]['curated_relation'] else '0') ==
+            json_to_marc['701'][0]['y'])
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_hepnames.py]: IMO you can avoid these.

File: tests/unit/dojson/fixtures/test_hepnames_record.xml [hunk 9]
```diff
@@ -127,6 +127,8 @@
         <subfield code="a">Fuller, George M.</subfield>
         <subfield code="g">PHD</subfield>
         <subfield code="i">INSPIRE-00147892</subfield>
+        <subfield code="x">1009206</subfield>
+        <subfield code="y">1</subfield>
```
Comment (jacquerie) [inline tests/unit/dojson/fixtures/test_hepnames_record.xml]: And this.

File: tests/unit/dojson/test_dojson_jobs.py [hunk 10]
```diff
@@ -230,39 +230,121 @@ def test_regions_from_043__a_splits_on_commas():
 
 
 def test_experiments_from_693__e():
+
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_jobs.py]: Leftover.

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 2 above)
Comment (tomaszgy) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Just found http://inspirehep.net/record/1019112/export/xme and it seems that `y` disappears if there's no link via recid or INSPIRE-ID.

File: tests/unit/dojson/test_dojson_jobs.py [hunk 11]
```diff
@@ -230,39 +230,121 @@ def test_regions_from_043__a_splits_on_commas():
...
     snippet = (
-        '<datafield tag="693" ind1=" " ind2=" ">'
-        '  <subfield code="e">CERN-LHC-ATLAS</subfield>'
-        '</datafield>'
-    )  # record/1471772
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_jobs.py]: Uhm? What was wrong with this guy?

File: tests/unit/dojson/test_dojson_jobs.py [hunk 12]
```diff
@@ -230,39 +230,121 @@ def test_regions_from_043__a_splits_on_commas():
...
-        '    <subfield code="e">CERN-LHC-CMS</subfield>'
-        '  </datafield>'
-        '  <datafield tag="693" ind1=" " ind2=" ">'
-        '    <subfield code="e">CALICE</subfield>'
-        '  </datafield>'
-        '  <datafield tag="693" ind1=" " ind2=" ">'
-        '    <subfield code="e">ILC</subfield>'
-        '  </datafield>'
+        '   <datafield tag="693" ind1=" " ind2=" ">'
+        '       <subfield code="e">CERN-LHC-ATLAS</subfield>'
```
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_jobs.py]: Why the switch to four spaces of indentation?

File: inspirehep/modules/records/jsonschemas/records/experiments.json (hunk 4 above)
Comment (tomaszgy) [inline inspirehep/modules/records/jsonschemas/records/experiments.json]: Agreed on that :smile:

**Comment by jacquerie** [comment]
In general it's a :+1:, but there's a few nits that should be addressed.

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 13]
```diff
@@ -450,11 +451,15 @@ def advisors(self, key, value):
 
     _degree_type = force_single_element(value.get('g'))
     degree_type = DEGREE_TYPES_MAP.get(_degree_type, 'Other')
+    recid = force_single_element(value.get('x'))
```
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: A space before this to separate the two different things the two pairs of lines are doing.

**Comment by jacquerie** [comment]
Oh, and please drop
```
* Updates JSON schemas.

* Includes tests.
```
from the commit message, because they add nothing.

File: tests/unit/dojson/test_dojson_jobs.py (hunk 11 above)
Comment (tomaszgy) [inline tests/unit/dojson/test_dojson_jobs.py]: If you go to old record [in the legacy](http://inspirehep.net/record/1471772/export/xme) now it contains also `0` subfield:

`<datafield tag="693" ind1=" " ind2=" ">`
`<subfield code="e">CERN-LHC-ATLAS</subfield>`
`<subfield code="0">1108541</subfield>`
`</datafield>`

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 2 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Uh? But this record had no `701__y` field... Do you have an example of an Hepname with a `701__y`?

**Pull request closed without merging**